store/src/java-test/com/zimbra/cs/datasource/imap/RemoteFolderTest.java

```diff
@@ -0,0 +1,136 @@
+        PowerMock.expectLastCall();
+
+        PowerMock.replay(mockConnection);
+
+        f.renameTo(newPath);
+
+        PowerMock.verify(mockConnection);
+    }
+
+    // Since the folder name's are same hence rename method will not be called
+    @Test
+    public void testRenameToGrandChildSameName() throws Exception {
+        final String originalPath = "Test/zimbra/ab";
+        final String newPath = "Test/Zimbra/ab";
+
+        ImapConnection mockConnection = EasyMock.createMock(ImapConnection.class);
+        RemoteFolder f = new RemoteFolder(mockConnection, originalPath);
+
+        PowerMock.replay(mockConnection);
+
+        f.renameTo(newPath);
+
+        PowerMock.verify(mockConnection);
+    }
+
+    // If we update the folder from Test/zimbra/pq to Test/Zimbra_1/pq then the
+    // rename of middle layer folder happens only when loop iterates to Test/Zimbra
+    // Below scenario run's in second iteration
+    @Test
+    public void testRenameToMiddleLayerFolder() throws Exception {
+        final String originalPath = "Test/zimbra";
+        final String newPath = "Test/Zimbra_1";
+
+        ImapConnection mockConnection = EasyMock.createMock(ImapConnection.class);
+        RemoteFolder f = new RemoteFolder(mockConnection, originalPath);
+
+        mockConnection.rename(originalPath, newPath);
+        PowerMock.expectLastCall();
+
+        PowerMock.replay(mockConnection);
+
+        f.renameTo(newPath);
+
+        PowerMock.verify(mockConnection);
+    }
+
+    // In the below test case since grand child name's are same hence rename action
+    // won't be called
+    @Test
+    public void testRenameToDifferentMiddleFolderNameandSameGrandChildName() throws Exception {
+        final String originalPath = "Test/zimbra/pq";
+        final String newPath = "Test/Zimbra_1/pq";
+
+        ImapConnection mockConnection = EasyMock.createMock(ImapConnection.class);
+        RemoteFolder f = new RemoteFolder(mockConnection, originalPath);
+
+        PowerMock.replay(mockConnection);
+
+        f.renameTo(newPath);
+
+        PowerMock.verify(mockConnection);
+    }
+
+    @Test
+    public void testRenameToDifferentGrandChildFolderName() throws Exception {
+        final String originalPath = "Test/zimbra/pq";
+        final String newPath = "Test/zimbra/pq_1";
+
+        ImapConnection mockConnection = EasyMock.createMock(ImapConnection.class);
+        RemoteFolder f = new RemoteFolder(mockConnection, originalPath);
+
+        mockConnection.rename(originalPath, newPath);
+        PowerMock.expectLastCall();
+
+        PowerMock.replay(mockConnection);
+
+        f.renameTo(newPath);
+
+        PowerMock.verify(mockConnection);
+    }
+
+    // In this case when middle layer and grand child folder name's are renamed at a
+    // time in this case in
+    // rename method newer path is modified in such a way that only "pq" is renamed
+    // in this loop
+    // an in next zoom Test/Zimbra get's updated to Test/zimbra_1 ref:
```

Comment:
`Test/Zimbra` is `Test/zimbra`, correct?

Reply:
Thanks , it was typo. Fixed it